src/Files.App/Services/Windows/WindowsRecentItemsService.cs

```diff
@@ -0,0 +1,318 @@
+				while (pEnumShellItems.Get()->Next(1, pShellItem.GetAddressOf()) == HRESULT.S_OK)
+				{
+					// Get top 20 items
+					if (index is 20)
+						break;
+
+					// Exclude folders
+					if (pShellItem.Get()->GetAttributes(SFGAO_FLAGS.SFGAO_FOLDER, out var attribute) == HRESULT.S_OK &&
+						(attribute & SFGAO_FLAGS.SFGAO_FOLDER) == SFGAO_FLAGS.SFGAO_FOLDER)
+						continue;
+
+					// Get the target path
+					pShellItem.Get()->GetDisplayName(SIGDN.SIGDN_DESKTOPABSOLUTEEDITING, out var szDisplayName);
+					var targetPath = szDisplayName.ToString();
+					PInvoke.CoTaskMemFree(szDisplayName.Value);
+
+					// Get the display name
+					pShellItem.Get()->GetDisplayName(SIGDN.SIGDN_NORMALDISPLAY, out szDisplayName);
+					var fileName = szDisplayName.ToString();
+					PInvoke.CoTaskMemFree(szDisplayName.Value);
+
+					// Strip the file extension except when the file name only contains extension (e.g. ".gitignore")
+					if (!FoldersSettingsService.ShowFileExtensions &&
+						SystemIO.Path.GetFileNameWithoutExtension(fileName) is string fileNameWithoutExtension)
+						fileName = string.IsNullOrEmpty(fileNameWithoutExtension) ? SystemIO.Path.GetFileName(fileName) : fileNameWithoutExtension;
+
+					// Get the date last modified
+					var shellItem2Iid = typeof(IShellItem2).GUID;
+					using ComPtr<IShellItem2> pShellItem2 = default;
+					hr = pShellItem.Get()->QueryInterface(&shellItem2Iid, (void**)pShellItem2.GetAddressOf());
+					hr = PInvoke.PSGetPropertyKeyFromName("System.DateModified", out var propertyKey);
+					hr = pShellItem2.Get()->GetString(propertyKey, out var szPropertyValue);
+					if (DateTime.TryParse(szPropertyValue.ToString(), out var lastModified))
+						lastModified = DateTime.MinValue;
+
+					recentItems.Add(new()
+					{
+						Path = targetPath,
+						Name = fileName,
+						ShellItem = pShellItem,
+						LastModified = lastModified,
+					});
+
+					index++;
+				}
+
+				if (recentItems.Count is 0)
+					return false;
+
+				var snapshot = isFolder ? RecentFolders : RecentFiles;
+
+				if (isFolder)
+				{
+					lock (_RecentFolders)
+					{
+						_RecentFolders.Clear();
+						_RecentFolders.AddRange(orderedRecentItems);
+					}
+				}
+				else
+				{
+					lock (_RecentFiles)
+					{
+						_RecentFiles.Clear();
+						_RecentFiles.AddRange(orderedRecentItems);
+					}
+				}
+
+				var eventArgs = GetChangedActionEventArgs(snapshot, orderedRecentItems);
+
+				if (isFolder)
+ 					RecentFoldersChanged?.Invoke(this, eventArgs);
+				else
+ 					RecentFilesChanged?.Invoke(this, eventArgs);
+
+				return true;
+			}
+			catch
+			{
+				return false;
+			}
+		}
+
+		private void Watcher_Changed(object sender, SystemIO.FileSystemEventArgs e)
+		{
+			_ = UpdateRecentFilesAsync();
+			_ = UpdateRecentFoldersAsync();
+		}
+
+		private NotifyCollectionChangedEventArgs GetChangedActionEventArgs(IReadOnlyList<RecentItem> oldItems, IList<RecentItem> newItems)
+		{
+			if (oldItems.Count - newItems.Count is 1)
```

Comment:
I think these two (add / remove) cases are inverted.

e.g. `oldItems = 7` and `newItems = 6` --> that means a deletion happened

Reply:
Oh my bad, let me fix shorty.

Reply:
Fixed